de.metas.banking/de.metas.banking.base/src/main/java/de/metas/banking/service/impl/BankStatementBL.java

```diff
@@ -97,16 +111,133 @@ public void handleAfterComplete(final I_C_BankStatement bankStatement)
+	public Optional<PaymentId> setOrCreateAndLinkPaymentToBankStatementLine(@NonNull final I_C_BankStatementLine line, @Nullable final PaymentId paymentIdToSet)
+	{
+		// a payment is already linked
+		if (line.getC_Payment_ID() > 0)
+		{
+			return Optional.of(PaymentId.ofRepoId(line.getC_Payment_ID()));
+		}
+
+		if (paymentIdToSet != null)
+		{
+			line.setC_Payment_ID(paymentIdToSet.getRepoId());
+			InterfaceWrapperHelper.save(line);
+			return Optional.of(paymentIdToSet);
+		}
+
+		if (line.getC_BPartner_ID() <= 0)
+		{
+			return Optional.empty();
+		}
+
+		final CurrencyId currencyId = CurrencyId.ofRepoId(line.getC_Currency_ID());
+		final BPartnerId bpartnerId = BPartnerId.ofRepoId(line.getC_BPartner_ID());
+		final OrgId orgId = OrgId.ofRepoId(line.getAD_Org_ID());
+		final LocalDate statementLineDate = TimeUtil.asLocalDate(line.getStatementLineDate());
+
+		final Optional<BankAccountId> bankAccountIdOptional = Services.get(IBPBankAccountDAO.class).retrieveFirstIdByBPartnerAndCurrency(bpartnerId, currencyId);
+		if (!bankAccountIdOptional.isPresent())
+		{
+			return Optional.empty();
+		}
+
+		final boolean isReceipt = line.getStmtAmt().signum() >= 0;
+		final BigDecimal payAmount = isReceipt ? line.getStmtAmt() : line.getStmtAmt().negate();
+
+		final PaymentId createdPaymentId = createAndCompletePayment(bankAccountIdOptional.get(), statementLineDate, payAmount, isReceipt, orgId, bpartnerId, currencyId);
+		line.setC_Payment_ID(createdPaymentId.getRepoId());
+		InterfaceWrapperHelper.save(line);
+		return Optional.of(createdPaymentId);
+	}
+
+	private void findAndLinkPaymentToBankStatementLineIfPossible(final I_C_BankStatementLine line)
+	{
+		// a payment is already linked
+		if (line.getC_Payment_ID() > 0)
+		{
+			return;
+		}
+
+		final boolean isReceipt = line.getStmtAmt().signum() >= 0;
+		final BigDecimal expectedPaymentAmount = isReceipt ? line.getStmtAmt() : line.getStmtAmt().negate();
+
+		final ImmutableSet<PaymentId> possiblePayments = Services.get(IPaymentDAO.class).retrieveAllMatchingPayments(isReceipt, expectedPaymentAmount, CurrencyId.ofRepoId(line.getC_Currency_ID()), BPartnerId.ofRepoId(line.getC_BPartner_ID()));
+
+		// don't create a new Payment and don't link any of the payments.
+		// the user must fix this case manually
+		if (possiblePayments.size() > 1)
+		{
+			return;
+		}
+
+		if (possiblePayments.size() == 1)
+		{
+			line.setC_Payment_ID(possiblePayments.iterator().next().getRepoId());
+			InterfaceWrapperHelper.save(line);
+		}
+	}
+
+	private PaymentId createAndCompletePayment(
+			@NonNull final BankAccountId bankAccountId,
+			@NonNull final LocalDate dateAcct,
+			@NonNull final BigDecimal payAmt,
+			final boolean isReceipt,
+			@NonNull final OrgId adOrgId,
+			@NonNull final BPartnerId bpartnerId,
+			@NonNull final CurrencyId currencyId)
+	{
+		final DefaultPaymentBuilder paymentBuilder;
+
+		if (isReceipt)
+		{
+			paymentBuilder = Services.get(IPaymentBL.class).newInboundReceiptBuilder();
```

Comment:
extract the IPaymentBL at least to a local variable

Reply:
done